hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/config/validation/internal/EmulatesMapValidator.java

```diff
@@ -0,0 +1,118 @@
+
+/**
+ * A {@link ConfigValidator} that validates that any property annotated with {@link EmulatesMap}. See {@link
+ * EmulatesMap} for more details.
+ */
+public class EmulatesMapValidator implements ConfigValidator {
+
+    @Override
+    @NonNull
+    public Stream<ConfigViolation> validate(@NonNull final Configuration configuration) {
+        Objects.requireNonNull(configuration, "Configuration cannot be null");
+        return ConfigReflectionUtils.getAllMatchingPropertiesForConstraintAnnotation(EmulatesMap.class, configuration)
+                .stream()
+                .map(this::convert)
+                .filter(Optional::isPresent)
+                .map(Optional::get);
+    }
+
+    @NonNull
+    private Optional<ConfigViolation> convert(@NonNull final AnnotatedProperty<EmulatesMap, ?> property) {
+        Objects.requireNonNull(property, "property cannot be null");
+        final Class<?> type = property.propertyType();
+        if (!List.class.isAssignableFrom(type) && !Set.class.isAssignableFrom(type)) {
+            final ConfigViolation violation =
+                    create(property, "Property is annotated with @EmulatesMap but is not a Set or List");
+            return Optional.of(violation);
+        }
+        final Type genericType = property.component().getGenericType();
+        if (genericType instanceof ParameterizedType parameterizedType) {
+            final Type singleGenericTypeArgument =
+                    ConfigReflectionUtils.getSingleGenericTypeArgument(parameterizedType);
+            if (!KeyValuePair.class.equals(singleGenericTypeArgument)) {
+                final ConfigViolation violation = create(
+                        property, "Property is annotated with @EmulatesMap but is not a Collection of KeyValuePair");
+                return Optional.of(violation);
+            }
+        } else {
+            final ConfigViolation violation =
+                    create(property, "Property is annotated with @EmulatesMap but is not a Collection of KeyValuePair");
+            return Optional.of(violation);
+        }
+        final var typedValue = (AnnotatedProperty<EmulatesMap, Collection<KeyValuePair>>) property;
+
+        final int uniqueKeyCount = typedValue.propertyValue().stream()
+                .map(KeyValuePair::key)
+                .collect(Collectors.toSet())
+                .size();
+        if (uniqueKeyCount != typedValue.propertyValue().size()) {
+            final ConfigViolation violation = create(typedValue, "Property contains duplicate keys");
+            return Optional.of(violation);
+        }
+        return Optional.empty();
+    }
+
+    @NonNull
+    private ConfigViolation create(
+            @NonNull final AnnotatedProperty<EmulatesMap, ?> property, @NonNull final String message) {
+        Objects.requireNonNull(property, "AnnotatedProperty cannot be null");
+        Objects.requireNonNull(message, "Message cannot be null");
+        return new ConfigViolation() {
```

Comment:
Please refresh my memory. 🙂 Why do you use an anonymous class here? I am pretty sure, there is a good reason, but I cannot remember it.